src/compiler/typescript_to_rust/generate_rust_canister.ts

```diff
@@ -47,13 +49,57 @@ function runAzleGenerate(
     canisterPath: string,
     canisterConfig: JSCanisterConfig
 ): Result<null, SpawnSyncError> {
-    // TODO to cut the time in half here, if we can just generate the binary
-    // TODO and then after that use the binary, that could work
-    // TODO the problem is that this binary is stored in a global target dir
-    // TODO we might need to have an azle-version-specific dir for the azle_generate binary
-    // TODO maybe that's fine and easy though
+    const azleGenerateBinPath = join(
+        GLOBAL_AZLE_CONFIG_DIR,
+        'azle_generate',
+        azleVersion,
+        'azle_generate'
+    );
+    const azleGenerateBinPathDebug = join(
+        GLOBAL_AZLE_TARGET_DIR,
+        'debug',
+        'azle_generate'
+    );
+    const azleGenerateDir = join(
+        canisterPath,
+        canisterConfig.root,
+        'azle_generate'
+    );
+
+    const shouldRebuild =
+        !existsSync(azleGenerateBinPath) || process.env.AZLE_REBUILD === 'true';
+
+    if (shouldRebuild) {
+        const buildResult = buildBinary(
+            azleGenerateBinPath,
+            azleGenerateBinPathDebug,
+            azleGenerateDir
+        );
+
+        if (buildResult.Err !== undefined) {
+            return Result.Err(buildResult.Err);
+        }
+    }
+
+    return runBinary(
+        azleGenerateBinPath,
+        azleGenerateDir,
+        compilerInfoPath,
+        canisterConfig
+    );
+}
+
+function buildBinary(
```

Comment:
Although this works, it's not intuitive to me. This method builds the binary, and copies it to the correct location. Maybe we should separate those two pieces out? or rename for clarity?

It's also interesting to me that  on line 97 we create the parent directories just from a bunch of parts of the path, but the azleGenerateBinPath theoretically could be somewhere else. We should somehow tie those two together.

Something like:
mkdirSync(path.dirname(azleGenerateBinPath)).

Then I would also just change the names on all this stuff on the inside of this function. So my sig would be:

```ts
function buildBinary(cmdDir: string, src: string, dest: string): Result<null, SpawnSyncError>
```

Like, we just need to know we're copying from src to dest, not what those paths are.

I just have a hard time keeping track of which path is which when all their names are all so similar.

Reply:
Great thoughts for the future